### DISCUSSION TITLE
Nest.js: `OnApplicationBootstrap()`: ValidationError: Using global EntityManager instance methods for context specific actions is disallowed.

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2536] [Q&A] [answered]
This seems to be related to #696, but there Nest.js is not used. I also use Fastify instead of Express/Apollo.

---

I need to initialise the app after it is started. That included building up some cache from PostgreSQL database. Therefore I created `InitAppService` in `app.service.ts` and added it to `providers` of `AppModule` (top-level module).

In the constructor of `InitAppService`, I added a service (e.g. `OtherService`) from another module (imported in `AppModule`, where I also have `MikroOrmModule.forRoot()`). Until now, the app starts as expected.

However, when I run run a method from the `OtherService` (e.g. `await this.otherService.whatever()`), which uses `createQueryBuilder()` on a repository, it does not work and throws the following error:

```c
ValidationError: Using global EntityManager instance methods for context specific actions is disallowed. If you need to work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead.
    at Function.cannotUseGlobalContext ({repo_root}/node_modules/@mikro-orm/core/dist/errors.js:79:16)
    at SqlEntityManager.getContext ({repo_root}/node_modules/@mikro-orm/core/dist/EntityManager.js:618:44)
    at SqlEntityRepository.get em [as em] ({repo_root}/node_modules/@mikro-orm/knex/dist/SqlEntityRepository.js:30:25)
    at SqlEntityRepository.createQueryBuilder ({repo_root}/node_modules/@mikro-orm/knex/dist/SqlEntityRepository.js:15:21)
    at MachineService.getMonitoredMachines ({repo_root}/src/machine/machine.service.ts:27:37)
    at InitAppService.onApplicationBootstrap ({repo_root}/src/app.service.ts:13:55)
    at MapIterator.iteratee ({repo_root}/node_modules/@nestjs/core/hooks/on-app-bootstrap.hook.js:22:43)
    at MapIterator.next ({repo_root}/node_modules/iterare/src/map.ts:9:39)
    at IteratorWithOperators.next ({repo_root}/node_modules/iterare/src/iterate.ts:19:28)
    at Function.from (<anonymous>) {
  entity: undefined
}
```

Do I really need to add instances of all needed repositories, `orm`, entities in order to make it work? Or should I rather move it into a separate module?

**Comment by B4nan** [comment, marked as answer]
> Do I really need to add instances of all needed repositories, orm, entities in order to make it work? Or should I rather move it into a separate module?

Not sure what you mean by that. That error says what you need to do, either allow the global context, or preferably work with a fork. Here it is quite obvious you are not inside request handlers, so no middlewares are fired, hence no context is created automatically - but you can still create it, either via `em.fork()` explicitly, or via `RequestContext.createAsync()`, or via `@UseRequestContext()` decorator. All of that should work, the last one is probably the easiest way for your use case. You should use that decorator on the top level method, where you know it will be fired only from the app startup (e.g. not a shared service method that would be used also from request handlers).

https://mikro-orm.io/docs/usage-with-nestjs/#request-scoped-handlers-in-queues

**Reply by tukusejssirs** [reply]
`@UseRequestContext()` works as expected. Thanks! :heart:

Sorry that you need to tell me too many information more than once because of me being a bad learner. Thanks for your patience! :pray:

**Reply by tukusejssirs** [reply]
@B4nan, how about a shared service method that would be used also from request handlers? I understand that it is bad idea to use `@UseRequestContext()` in that case.

Should I create a wrapper function which would be decorated with `@UseRequestContext()`, in which I would call the main function, which could be used also from request handlers? Or is there a better approach?

**Reply by B4nan** [reply]
Maybe you could use the decorator directly in app module, not sure. You could always create a proxy method with the decorator, and use that from app bootstrap, while the raw method underneath would be used from regular request scoped handlers (or other services).

**Reply by tukusejssirs** [reply]
> Maybe you could use the decorator directly in app module, not sure.

I want not able to do so, it errored out with the original error.

> You could always create a proxy method with the decorator, and use that from app bootstrap, while the raw method underneath would be used from regular request scoped handlers (or other services).

That _proxy method_ sounds like the _wrapper method_ I suggested, am I right?
